[Jian Wei] iP

jianoway opened the pull request:
# InferiorDuke

Welcome, ***stalker***, to my *personal* rendition of ~~Duke~~ aka InferiorDuke, where I try to follow requirements for my iP in hopes of passing this module!

> "If it ain't broke, don't fix it" - Thomas Bertram Lance ([Source](https://digital.hagley.org/Nationbiz_197705#page/30/mode/2up))

# Features

As this is a work in progress, most of the features currently available are still being worked upon with many more planned for the future!

## Available Features ✔️

* Able to add 3 different types of tasks (Event, Deadline and ToDo)
* View your Task list with the list command, with your tasks stored locally on your computer
* Mark tasks as done or undone with the done and undo commands
* Delete tasks from the Task list

## Upcoming Key Features 📝

1. Implement Graphical User Interface (GUI) 
2. Implement Gradle
3. Add JavaDocs to all public methods availble

## Additional Optionals

- [X] Add JavaDocs to all public methods
- [x] All code 100% CheckStyle
- [x] Make cute logo for InferiorDuke

# Code Quality

Hopefully, my code is in accordance to the quality acceptable by our tutors and professors. An sample would be as follows:

```java
public static void main(String[] args) {
    new Duke("data/duke.txt").run();
}
```

## Errors

One thing you might notice about my code is that I chose to create my errors as DukeErrors as seen below:

```java
    public class DukeException extends Exception
```

I used to use DukeErrors as an enumeration but I decided that making it extend the Exception class would make the flow of logic cleaner.

jianoway committed:
Add impatient echo bot

jianoway committed:
Add list feature

jianoway committed:
Add mark as done and undone to tasklist

jianoway committed:
Add ToDos, Events and Deadlines

jianoway committed:
Add some automated test cases

jianoway committed:
Add some error handling

jianoway committed:
Add some test cases for error handling

jianoway committed:
Add delete functionality

jianoway committed:
Add some test cases for the delete function

jianoway committed:
Add some form of enum...

jianoway committed:
Add local save

jianoway committed:
Add LocalDate support for Event and Deadline

jianoway committed:
Add Duke package

jianoway committed:
Add more OOP

jianoway committed:
Fix non-termination bug

jianoway committed:
Add JUnit Tests

jianoway committed:
Add JAR

jianoway committed:
Add search function

jianoway committed:
Add some standard to my code

jianoway committed:
Merge level 9 and coding standard branches

jianoway committed:
Resolve merge conflicts x2

jianoway committed:
Add some Javadocs

jianoway committed:
Update README.md

jianoway committed:
Add Gradle support

jianoway committed:
Merge branch 'branch-A-Gradle'

jianoway committed:
Add checkStyle support for Gradle

jianoway committed:
Modify DukeError to DukeException

jianoway committed:
Add GUI

limdanqi commented on src/main/java/duke/errors/DukeException.java:
```diff
@@ -0,0 +1,125 @@
+        EMPTY_SEARCH_STRING(14, "Please key in something for me to find!");
+
+        /** The error code. */
+        private final int code;
+
+        /** The error message. */
+        private final String description;
```
Perhaps code can be named as errorCode for better clarity, the same for description, getCode() and getDescription() methods below?

jianoway replied:
Thanks for the suggestion, will implement in my next commit! :)

limdanqi commented on src/main/java/duke/errors/DukeException.java:
```diff
@@ -0,0 +1,125 @@
+         * @param code The error code
+         * @return The respective DukeError
+         */
+        public static DukeExceptionType getError(int code) {
```
This method is fine but perhaps the method can be shorter for better readability? Maybe instead of having all the exceptions under DukeException, there could be multiple sub-classes?

jianoway replied:
Was thinking of that as well, but was afraid that having 13 and more subclasses would seem excessive. Will try to implement as subclasses!

limdanqi commented on src/main/java/duke/parser/CustomDateFormatter.java:
```diff
@@ -0,0 +1,93 @@
+            try {
+                return LocalDate.parse(text, DateTimeFormatter.ofPattern(pattern));
+            } catch (DateTimeParseException e) {
+                // Don't want to print anything here as it would disrupt the UI
```
Perhaps there could still be errors printed in the case whereby the date provided is invalid?

jianoway replied:
Missed this out, was thinking of the same thing after adapting the More-OOP structure but didn't get around to implementing in the UI. Thanks~

limdanqi commented on src/main/java/duke/parser/Cleaner.java:
```diff
@@ -0,0 +1,251 @@
+     */
+    private String listCleaner(String fullCommand) {
+        if (fullCommand.strip().length() >= 5) {
+            return "error 1";
```
Maybe a named constant for all the errors can be used instead as this is a case of magic string? Also, it can help those that are reading your code understand what the error is.

jianoway replied:
I didn't know what magic string was until you mentioned it, thanks for that haha learnt something new!

limdanqi commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,167 @@
+     */
+    public void editStorage(int lineNumber, CommandType commandType) {
+        int actualLineNumber = lineNumber + 1;
+        try {
```
The try block could be shorter by implementing more helper methods?

jianoway replied:
Ah yes, it does seem pretty long now that you've mentioned it, will fix it!

limdanqi reviewed (commented):
In general, I find your code to be pretty cool. However, the code can be more succinct so that it can appear to be cleaner and easier to read. Good job! :))

jianoway committed:
Restructure DukeExceptions and clean up code

Siddharth-Sid commented on src/main/java/duke/errors/DukeException.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Override
+    public String toString() {
```
I think in general we don't do use toString to get the exception details instead we use a method called getMessage which is a method of the Exception class. That said I think this is fine as well. Excellent work on the multiple types of exceptions you have created from this abstract class.

jianoway replied:
Thanks for the suggestion! Wasn't aware of using getMessage haha have implemented the change~ :)

Siddharth-Sid commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,44 @@
+        return str;
+    }
+
+    @Override
```
May I recommend that you call super.getDataString() instead

jianoway replied:
Thanks for the suggestion! Made my code much cleaner!! (:

Siddharth-Sid commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public String getDataString() {
```
May I recommend that you call super.getDataString() instead

Siddharth-Sid commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,59 @@
+    protected String description;
+    protected boolean isDone;
+    protected TaskType taskType;
+    enum TaskType {
```
Is this Enum needed? 
Also if needed it is better to declare it outside this class i.e. in a separate file

jianoway replied:
Only implemented the enum to fulfil the 'A-Enum' optional so I think I shall leave it in for now~ Will follow your suggestion to declare it outside the class~!

Siddharth-Sid commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,22 @@
+        super(description);
+        this.taskType = TaskType.T;
+    }
+
```
I recommend you Override the getDataString Method here and simply call the super.getDataString.

I know it may seem redundant but I think it is a better practice.

jianoway replied:
Cool! Will proceed to implement (:

Siddharth-Sid commented on text-ui-test/ACTUAL.TXT:
```diff
@@ -0,0 +1,41 @@
+  [D][ ] Submit proposal (by: 33 May)
+Noted. I've removed this task:
+  [T][X] Feed Cows
+Now you have 3 tasks in the list.
```
Not that this is needed now but if you are testing the Input you should also test all the possible error logs since it will help track if anything changes for the both the Happy Flow and the Error flow

jianoway replied:
To be honest, these test files were from like week 2/3 and are really outdated. Will be revamping it soon! Thanks for noticing!!

Siddharth-Sid reviewed (commented):
I think you have done an excellent job and there are little or not errors. I have made a couple of suggestions that may help you so request you to take a look.

jianoway committed:
Reorder LocalDate formatter pattern

jianoway committed:
Update README.md

jianoway committed:
Update README.md

jianoway committed:
Merge branch 'master' of https://github.com/jianoway/ip

jianoway committed:
Add some changes from PR

jianoway committed:
Merge branch 'Use-GFMD'

jianoway committed:
Latest release for Duke

The new release has integration with GUI and also a new JAR file.
This release also has implemented feedback from previous PR reviews.

Please have mercy!

jianoway committed:
Add new JAR file

jianoway committed:
Merge branch 'master' of https://github.com/jianoway/ip

jianoway committed:
Add some assertions

jianoway committed:
Improve code quality

jianoway committed:
Merge pull request #1 from jianoway/branch-A-Assertions

Add some assertions

jianoway committed:
Merge pull request #2 from jianoway/branch-A-CodeQuality

Improve code quality

jianoway committed:
Add sorting by date and task typeas well as SingleTimedTasks

jianoway committed:
Merge pull request #3 from jianoway/branch-C-Sort

Add sorting by date and task typeas well as SingleTimedTasks

jianoway committed:
Add some GUI changes

jianoway committed:
Add some more GUI modifications, namely a more pleasing background colour and smaller profile pictures

jianoway committed:
Merge pull request #4 from jianoway/A-BetterGui

Add some more GUI modifications, like a more pleasing background colour and smaller profile pictures

jianoway committed:
Add product image

jianoway committed:
Set theme jekyll-theme-minimal

jianoway committed:
Add the User Guide

jianoway committed:
Merge pull request #5 from jianoway/branch-A-UserGuide

Add the User Guide

jianoway committed:
Update README.md

jianoway committed:
Update README.md

jianoway committed:
Update README.md

jianoway committed:
Update README.md

jianoway committed:
Fix DialogBox truncating

jianoway committed:
Merge branch 'master' of https://github.com/jianoway/ip

jianoway committed:
Remove unnecessary data file

jianoway committed:
Clean up code

jianoway committed:
Add Assertion
